{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "hw2_Antonnikov.ipynb",
      "provenance": [],
      "collapsed_sections": [
        "mFvUKqx0Lss5"
      ]
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# Подготовка файлов и программ"
      ],
      "metadata": {
        "id": "mFvUKqx0Lss5"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Скачиваем файл scaffolds.fasta из первого дз "
      ],
      "metadata": {
        "id": "K23ThE-1JaQu"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "JYvnz2etG6xv",
        "outputId": "f7484cd8-3a6e-4066-e470-408bc7b2c28e"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2021-12-18 10:47:02--  https://raw.githubusercontent.com/greggasd/hse21_hw1/main/data/scaffolds.fasta\n",
            "Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n",
            "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 3961243 (3.8M) [text/plain]\n",
            "Saving to: ‘scaffolds.fasta’\n",
            "\n",
            "scaffolds.fasta     100%[===================>]   3.78M  --.-KB/s    in 0.07s   \n",
            "\n",
            "2021-12-18 10:47:02 (53.4 MB/s) - ‘scaffolds.fasta’ saved [3961243/3961243]\n",
            "\n"
          ]
        }
      ],
      "source": [
        "!wget -nc https://raw.githubusercontent.com/greggasd/hse21_hw1/main/data/scaffolds.fasta"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Скачиваем GeneMark и ключ от него\n",
        "http://opal.biology.gatech.edu/GeneMark/license_download.cgi\n",
        "\n",
        "Скачиваем по ссылке и кидаем в директорию колаба"
      ],
      "metadata": {
        "id": "U0n7QYGAJkIR"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!gzip -d gm_key_64.gz\n",
        "!tar -xzvf gms2_linux_64.tar.gz\n",
        "!cp -v gm_key_64 ~/.gmhmmp2_key"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "SJ1D_NtCHUYz",
        "outputId": "58e13683-17b0-4b90-ed8f-6784acbacb3f"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "gms2_linux_64/\n",
            "gms2_linux_64/gms2.pl\n",
            "gms2_linux_64/README_GMS2.txt\n",
            "gms2_linux_64/gmhmmp2_key\n",
            "gms2_linux_64/mgm_4.mod\n",
            "gms2_linux_64/gmhmmp2\n",
            "gms2_linux_64/INSTALL\n",
            "gms2_linux_64/mgm_11.mod\n",
            "gms2_linux_64/biogem\n",
            "gms2_linux_64/LICENSE\n",
            "gms2_linux_64/mgm_25.mod\n",
            "gms2_linux_64/compp\n",
            "gms2_linux_64/mgm_15.mod\n",
            "'gm_key_64' -> '/root/.gmhmmp2_key'\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Качаем данные по бактерии T.oleivorans"
      ],
      "metadata": {
        "id": "Br66yg9TKtUC"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!curl -fsSL ftp://ftp.ncbi.nlm.nih.gov/entrez/entrezdirect/install-edirect.sh | sh -\n",
        "!~/edirect/efetch -db nuccore -id HF680312 -format gb > T_oleivorans_MIL_1.gbk\n",
        "!~/edirect/efetch -db nuccore -id HF680312 -format gene_fasta > T_oleivorans_MIL_1.genes.fasta\n",
        "!~/edirect/efetch -db nuccore -id HF680312 -format fasta_cds_aa > T_oleivorans_MIL_1.proteins.fasta"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "wYShCViVKwky",
        "outputId": "c5221a89-5797-4ed3-cbd7-719711dce68b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Entrez Direct has been successfully downloaded and installed.\n",
            "\n",
            "In order to complete the configuration process, please execute the following:\n",
            "\n",
            "  echo \"export PATH=\\${PATH}:/root/edirect\" >> ${HOME}/.bashrc\n",
            "\n",
            "or manually edit the PATH variable assignment in your .bashrc file.\n",
            "\n",
            "Would you like to do that automatically now? [y/N]\n",
            "Holding off, then.\n",
            "\n",
            "To activate EDirect for this terminal session, please execute the following:\n",
            "\n",
            "export PATH=${PATH}:${HOME}/edirect\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Добавляем остальные прогрмаммы"
      ],
      "metadata": {
        "id": "jQ6Shf2NL0wZ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!apt update\n",
        "!apt install ncbi-blast+\n",
        "!wget https://github.com/shenwei356/seqkit/releases/download/v2.1.0/seqkit_linux_amd64.tar.gz\n",
        "!tar -xzvf seqkit_linux_amd64.tar.gz\n",
        "!chmod a+x seqkit"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7EtGBtcKMIVL",
        "outputId": "81f4c4fb-e5d0-4355-8972-25e79afbe819"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[33m\r0% [Working]\u001b[0m\r            \rGet:1 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ InRelease [3,626 B]\n",
            "Ign:2 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease\n",
            "Get:3 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic InRelease [15.9 kB]\n",
            "Ign:4 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  InRelease\n",
            "Get:5 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Release [696 B]\n",
            "Hit:6 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  Release\n",
            "Get:7 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Release.gpg [836 B]\n",
            "Get:8 http://security.ubuntu.com/ubuntu bionic-security InRelease [88.7 kB]\n",
            "Hit:9 http://archive.ubuntu.com/ubuntu bionic InRelease\n",
            "Get:10 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ Packages [73.9 kB]\n",
            "Get:11 http://archive.ubuntu.com/ubuntu bionic-updates InRelease [88.7 kB]\n",
            "Hit:12 http://ppa.launchpad.net/cran/libgit2/ubuntu bionic InRelease\n",
            "Get:14 http://ppa.launchpad.net/deadsnakes/ppa/ubuntu bionic InRelease [15.9 kB]\n",
            "Get:15 http://archive.ubuntu.com/ubuntu bionic-backports InRelease [74.6 kB]\n",
            "Get:16 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Packages [833 kB]\n",
            "Hit:17 http://ppa.launchpad.net/graphics-drivers/ppa/ubuntu bionic InRelease\n",
            "Get:18 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic/main Sources [1,820 kB]\n",
            "Get:19 http://security.ubuntu.com/ubuntu bionic-security/main amd64 Packages [2,461 kB]\n",
            "Get:20 http://archive.ubuntu.com/ubuntu bionic-updates/universe amd64 Packages [2,230 kB]\n",
            "Get:21 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic/main amd64 Packages [933 kB]\n",
            "Get:22 http://security.ubuntu.com/ubuntu bionic-security/universe amd64 Packages [1,452 kB]\n",
            "Get:23 http://security.ubuntu.com/ubuntu bionic-security/restricted amd64 Packages [691 kB]\n",
            "Get:24 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 Packages [2,898 kB]\n",
            "Get:25 http://archive.ubuntu.com/ubuntu bionic-backports/universe amd64 Packages [12.6 kB]\n",
            "Get:26 http://archive.ubuntu.com/ubuntu bionic-backports/main amd64 Packages [11.6 kB]\n",
            "Get:27 http://ppa.launchpad.net/deadsnakes/ppa/ubuntu bionic/main amd64 Packages [45.9 kB]\n",
            "Fetched 13.8 MB in 3s (4,023 kB/s)\n",
            "Reading package lists... Done\n",
            "Building dependency tree       \n",
            "Reading state information... Done\n",
            "79 packages can be upgraded. Run 'apt list --upgradable' to see them.\n",
            "Reading package lists... Done\n",
            "Building dependency tree       \n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  ncbi-data\n",
            "The following NEW packages will be installed:\n",
            "  ncbi-blast+ ncbi-data\n",
            "0 upgraded, 2 newly installed, 0 to remove and 79 not upgraded.\n",
            "Need to get 13.1 MB of archives.\n",
            "After this operation, 66.7 MB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu bionic/universe amd64 ncbi-data all 6.1.20170106-2 [3,645 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu bionic/universe amd64 ncbi-blast+ amd64 2.6.0-1 [9,446 kB]\n",
            "Fetched 13.1 MB in 2s (8,213 kB/s)\n",
            "Selecting previously unselected package ncbi-data.\n",
            "(Reading database ... 155222 files and directories currently installed.)\n",
            "Preparing to unpack .../ncbi-data_6.1.20170106-2_all.deb ...\n",
            "Unpacking ncbi-data (6.1.20170106-2) ...\n",
            "Selecting previously unselected package ncbi-blast+.\n",
            "Preparing to unpack .../ncbi-blast+_2.6.0-1_amd64.deb ...\n",
            "Unpacking ncbi-blast+ (2.6.0-1) ...\n",
            "Setting up ncbi-data (6.1.20170106-2) ...\n",
            "Setting up ncbi-blast+ (2.6.0-1) ...\n",
            "Processing triggers for hicolor-icon-theme (0.17-2) ...\n",
            "Processing triggers for man-db (2.8.3-2ubuntu0.1) ...\n",
            "--2021-12-18 10:57:34--  https://github.com/shenwei356/seqkit/releases/download/v2.1.0/seqkit_linux_amd64.tar.gz\n",
            "Resolving github.com (github.com)... 140.82.113.4\n",
            "Connecting to github.com (github.com)|140.82.113.4|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/52715040/ea4018ec-70e8-40ac-a4aa-3326b04e07c4?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20211218%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20211218T105734Z&X-Amz-Expires=300&X-Amz-Signature=69d0d2f5124258d9be0ff2cc19e0279b3a00f5673097eb2aa282d8f8f59355eb&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=52715040&response-content-disposition=attachment%3B%20filename%3Dseqkit_linux_amd64.tar.gz&response-content-type=application%2Foctet-stream [following]\n",
            "--2021-12-18 10:57:34--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/52715040/ea4018ec-70e8-40ac-a4aa-3326b04e07c4?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20211218%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20211218T105734Z&X-Amz-Expires=300&X-Amz-Signature=69d0d2f5124258d9be0ff2cc19e0279b3a00f5673097eb2aa282d8f8f59355eb&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=52715040&response-content-disposition=attachment%3B%20filename%3Dseqkit_linux_amd64.tar.gz&response-content-type=application%2Foctet-stream\n",
            "Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n",
            "Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.108.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 6650850 (6.3M) [application/octet-stream]\n",
            "Saving to: ‘seqkit_linux_amd64.tar.gz’\n",
            "\n",
            "seqkit_linux_amd64. 100%[===================>]   6.34M  --.-KB/s    in 0.08s   \n",
            "\n",
            "2021-12-18 10:57:34 (77.7 MB/s) - ‘seqkit_linux_amd64.tar.gz’ saved [6650850/6650850]\n",
            "\n",
            "seqkit\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Скачиваем БД белков SwissProt"
      ],
      "metadata": {
        "id": "DvXwwE7zf0Qw"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz\n",
        "!gzip -d uniprot_sprot.fasta.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "6nOSPsdVf2Lo",
        "outputId": "529ceb7e-290f-4ac2-bd4a-645cd1ba444d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2021-12-18 10:58:30--  https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz\n",
            "Resolving ftp.uniprot.org (ftp.uniprot.org)... 128.175.240.195\n",
            "Connecting to ftp.uniprot.org (ftp.uniprot.org)|128.175.240.195|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 90527596 (86M) [application/x-gzip]\n",
            "Saving to: ‘uniprot_sprot.fasta.gz’\n",
            "\n",
            "uniprot_sprot.fasta 100%[===================>]  86.33M  74.0MB/s    in 1.2s    \n",
            "\n",
            "2021-12-18 10:58:31 (74.0 MB/s) - ‘uniprot_sprot.fasta.gz’ saved [90527596/90527596]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Выполнение первого задания"
      ],
      "metadata": {
        "id": "PIBVoyeyf9UJ"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Первым делом ищем гены"
      ],
      "metadata": {
        "id": "GXcCRnv6Nyum"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!./gms2_linux_64/gms2.pl  --seq scaffolds.fasta  --genome-type bacteria  --fnn genes.fasta  --faa proteins.fasta"
      ],
      "metadata": {
        "id": "DOPW6OIYHz0C"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Далее определяем их назначение"
      ],
      "metadata": {
        "id": "QSwpykMdN2bu"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!makeblastdb -dbtype prot -in T_oleivorans_MIL_1.proteins.fasta -out T_oleivorans_MIL_1.proteins\n",
        "!blastp -query proteins.fasta -db T_oleivorans_MIL_1.proteins -evalue 1e-10 -outfmt 6 > scaffolds.hits_from_MIL_1.txt\n",
        "\n",
        "\n",
        "!cut -f 1 scaffolds.hits_from_MIL_1.txt | sort -n | uniq > proteins.with_hits_from_MIL_1.txt\n",
        "!./seqkit grep --invert-match -f proteins.with_hits_from_MIL_1.txt proteins.fasta -o proteins.without_MIL_1.fasta\n",
        "\n",
        "\n",
        "!makeblastdb -dbtype prot -in uniprot_sprot.fasta -out uniprot_sprot\n",
        "!blastp -query proteins.without_MIL_1.fasta -db uniprot_sprot -evalue 1e-10 -outfmt 6 > scaffolds.hits_from_SwissProt.txt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "QfLgPcQVFRmf",
        "outputId": "3e4ef993-2548-47f1-e352-76e554d1b4e5"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "\n",
            "Building a new DB, current time: 12/18/2021 11:01:08\n",
            "New DB name:   /content/T_oleivorans_MIL_1.proteins\n",
            "New DB title:  T_oleivorans_MIL_1.proteins.fasta\n",
            "Sequence type: Protein\n",
            "Keep MBits: T\n",
            "Maximum file size: 1000000000B\n",
            "Adding sequences from FASTA; added 3666 sequences in 0.098691 seconds.\n",
            "[INFO]\u001b[0m 3328 patterns loaded from file\n",
            "\n",
            "\n",
            "Building a new DB, current time: 12/18/2021 11:02:11\n",
            "New DB name:   /content/uniprot_sprot\n",
            "New DB title:  uniprot_sprot.fasta\n",
            "Sequence type: Protein\n",
            "Keep MBits: T\n",
            "Maximum file size: 1000000000B\n",
            "Adding sequences from FASTA; added 565928 sequences in 12.9599 seconds.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "И теперь считаем финальную статистику"
      ],
      "metadata": {
        "id": "VdoQIFL0SFH3"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "total_genes = !grep -c '>' genes.fasta\n",
        "total_genes = int(total_genes[0])\n",
        "\n",
        "rest_genes = !grep -c '>' proteins.without_MIL_1.fasta\n",
        "rest_genes = int(rest_genes[0])\n",
        "\n",
        "swiss_genes = !cut -f 1 scaffolds.hits_from_SwissProt.txt | sort | uniq | wc -l\n",
        "swiss_genes = int(swiss_genes[0])\n",
        "\n",
        "print(\"Cколько было предсказано генов всего?\", total_genes)\n",
        "print(\"Cколько из них удалось аннотировать с помощью сравнения с бактерией?\", total_genes-rest_genes)\n",
        "print(\"Сколько с помощью БД SwissProt?\", swiss_genes)\n",
        "print(\"Kакое кол-во белков остались без аннотации функции?\", rest_genes-swiss_genes)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "nfNKGMRbSNfg",
        "outputId": "bf5bbda9-126b-4aea-f87b-7ad6520c0abf"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Cколько было предсказано генов всего? 3606\n",
            "Cколько из них удалось аннотировать с помощью сравнения с бактерией? 3328\n",
            "Сколько с помощью БД SwissProt? 52\n",
            "Kакое кол-во белков остались без аннотации функции? 226\n"
          ]
        }
      ]
    }
  ]
}